src/infra/core/src/repos/metadata_chain_impl.rs

```diff
@@ -492,82 +413,24 @@ where
         Ok(())
     }
 
-    async fn validate_append_watermark_is_monotonic(
-        &self,
-        new_block: &MetadataBlock,
+    async fn accept_append_validators<'a>(
+        &'a self,
+        decisions: &'a mut [Decision],
+        visitors: &mut [&mut dyn MetadataChainVisitor<Error = AppendError>],
+        prev_append_block_hash: Option<&'a Multihash>,
     ) -> Result<(), AppendError> {
-        if let Some(new_block) = new_block.as_data_stream_block() {
-            let prev_wm = self
-                .iter_blocks_interval(new_block.prev_block_hash.unwrap(), None, false)
-                .filter_data_stream_blocks()
-                .map_ok(|(_, b)| b.event.new_watermark)
-                .try_first()
-                .await
-                .int_err()?;
-
-            if let Some(prev_wm) = &prev_wm {
-                match (prev_wm, new_block.event.new_watermark) {
-                    (Some(_), None) => {
-                        return Err(AppendValidationError::WatermarkIsNotMonotonic.into())
-                    }
-                    (Some(prev_wm), Some(new_wm)) if prev_wm > new_wm => {
-                        return Err(AppendValidationError::WatermarkIsNotMonotonic.into())
-                    }
-                    _ => (),
-                }
-            }
-        }
-        Ok(())
-    }
+        let have_already_stopped = decisions.iter().all(|decision| *decision == Decision::Stop);
```

Comment:
Minor: maybe holding a counter of active visitors among with `decisions` slice could reduce this comparison to 1 operation, on the other hand we would have to detect changes in `decisions` within the traversal code.

Reply:
Moved to https://github.com/kamu-data/kamu-cli/pull/543#discussion_r1517879457